src/htm/algorithms/Connections.cpp

```diff
@@ -160,19 +189,36 @@ Synapse Connections::createSynapse(Segment segment,
 }
 
 bool Connections::segmentExists_(const Segment segment) const {
-  NTA_CHECK(segment < segments_.size());
+  if(segment >= segments_.size()) return false; //OOB segment
+
   const SegmentData &segmentData = segments_[segment];
   const vector<Segment> &segmentsOnCell = cells_[segmentData.cell].segments;
-  return (std::find(segmentsOnCell.cbegin(), segmentsOnCell.cend(), segment) !=
-          segmentsOnCell.cend());
+  return (std::find(segmentsOnCell.cbegin(), segmentsOnCell.cend(), segment) != segmentsOnCell.cend()); //TODO if too slow, also create "fast" variant, as synapseExists_()
 }
 
-bool Connections::synapseExists_(const Synapse synapse) const {
+bool Connections::synapseExists_(const Synapse synapse, bool fast) const {
+  if(synapse >= synapses_.size()) return false; //out of bounds. Can happen after serialization, where only existing synapses are stored.
+
+#ifdef NTA_ASSERTIONS_ON
+  fast = false; //in Debug, do the proper, slow check always
+#endif
+  if(!fast) {
+  //proper but slow method to check for valid, existing synapse
   const SynapseData &synapseData = synapses_[synapse];
   const vector<Synapse> &synapsesOnSegment =
       segments_[synapseData.segment].synapses;
-  return (std::find(synapsesOnSegment.begin(), synapsesOnSegment.end(),
-                    synapse) != synapsesOnSegment.end());
+  const bool found = (std::find(synapsesOnSegment.begin(), synapsesOnSegment.end(), synapse) != synapsesOnSegment.end());
+  //validate the fast & slow methods for same result:
+#ifdef NTA_ASSERTIONS_ON
+  const bool removed = synapses_[synapse].permanence == -1;
+  NTA_ASSERT( (removed and not found) or (not removed and found) );
+#endif
+  return found;
+
+  } else {
+  //quick method. Relies on hack in destroySynapse() where we set synapseData.permanence == -1
+  return synapses_[synapse].permanence != -1;
```

Comment:
this code above is quite slow, but is needed and needed in a heavily used `dataForSynapse()`, therefore I created this "fast" variant/hack.